### PR TITLE
stm32: dma: enable dma with peripheral

### LINKED ISSUE #20825: stm32: dma: enable dma with peripheral using DMAMUX
**Is your enhancement proposal related to a problem? Please describe.**
On some STM32 series, a DMAMUX IP is present to handle the peripheral DMA channels multiplexing. This new IP is not supported yet on STM32 DMA driver

**Describe the solution you'd like**
So to support use of DMA by peripherals on these series, stm32 dma driver should be updated 
to support channel configuration using DMAMUX

### PR DESCRIPTION
**Pull request by FRASTM**
Test application and example configuring the spi  peripheral as a client of the dma.
The peripheral request is defined by the client including the channels are fixed for transfer. 

The DT is filled with parameters to set the peripheral request on dma
The dma  channels used for the client transfer (rx, tx) are assigned by the DT among the total nb of channels.
Also tested on nucleo_l476rg board, where the dma requests are fixed (no multiplexer)
Also tested on nucleo_f411re board, where the dma V1 requests are fixed (no multiplexer)

Requires https://github.com/zephyrproject-rtos/zephyr/pull/23157

Fix https://github.com/zephyrproject-rtos/zephyr/issues/20825

Signed-off-by: Francois Ramu francois.ramu@st.com

### CODE REVIEW

**Comment by FRASTM** [comment]
rebase

**Comment by FRASTM** [comment]
changed the DTS for dma client of the SPI for nucleo_l476
removed `compatible = "st,stm32-spi"; `  in the .overlay for this test

**Comment by FRASTM** [comment]
changed the dts_fixup.h for l4 on DT_FLASH_DEV_NAME

**Comment by FRASTM** [comment]
rebase

File: soc/arm/st_stm32/stm32l4/dts_fixup.h [hunk 1]
```diff
@@ -30,4 +30,24 @@
...
+
+#define DT_SPI_1_DMA_CONTROLLER_RX		\
+			DT_ST_STM32_SPI_40013000_RX_DMAS_CONTROLLER
+#define DT_SPI_1_DMA_CHANNEL_RX	DT_ST_STM32_SPI_40013000_RX_DMAS_CHANNEL
+#define DT_SPI_1_DMA_SLOT_RX	DT_ST_STM32_SPI_40013000_RX_DMAS_SLOT
+#define DT_SPI_1_DMA_CHANNEL_CONFIG_RX		\
+			DT_ST_STM32_SPI_40013000_RX_DMAS_CHANNEL_CONFIG
+#define DT_SPI_1_DMA_FEATURES_RX		\
+			DT_ST_STM32_SPI_40013000_RX_DMAS_FEATURES
+#endif /* CONFIG_SPI_STM32_DMA */
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l4/dts_fixup.h:51]: This should not be needed anymore

File: soc/arm/st_stm32/stm32f4/dts_fixup.h [hunk 2]
```diff
@@ -176,4 +176,24 @@
...
+#define DT_SPI_1_DMA_CONTROLLER_RX		\
+			DT_ST_STM32_SPI_40013000_RX_DMAS_CONTROLLER
+#define DT_SPI_1_DMA_CHANNEL_RX	DT_ST_STM32_SPI_40013000_RX_DMAS_CHANNEL
+#define DT_SPI_1_DMA_SLOT_RX	DT_ST_STM32_SPI_40013000_RX_DMAS_SLOT
+#define DT_SPI_1_DMA_CHANNEL_CONFIG_RX		\
+			DT_ST_STM32_SPI_40013000_RX_DMAS_CHANNEL_CONFIG
+#define DT_SPI_1_DMA_FEATURES_RX		\
+			DT_ST_STM32_SPI_40013000_RX_DMAS_FEATURES
+#endif /* CONFIG_SPI_STM32_DMA */
+
```
Comment (erwango) [inline soc/arm/st_stm32/stm32f4/dts_fixup.h:198]: Not needed anymore

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/spi/spi_ll_stm32.c [hunk 3]
```diff
@@ -59,7 +59,68 @@ u32_t nop_tx;
 /* This function is executed in the interrupt context */
 static void dma_callback(void *arg, u32_t channel, int status)
 {
-	/* TODO: callback function where arg directly holds the client data */
+	/* callback_arg directly holds the client data */
```
Comment (erwango) [inline drivers/spi/spi_ll_stm32.c:62]: Can this be squashed to previous commit ?

File: dts/bindings/dma/dma.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
+# Copyright (c) 2019, Linaro Limited
+# Copyright (c) 2019, Song Qiang <songqiang1304521@gmail.com>
+# SPDX-License-Identifier: Apache-2.0
+
+# Common fields for DMA users
```
Comment (erwango) [inline dts/bindings/dma/dma.yaml:5]: This binding is not needed anymore.
Properties moved to base.yaml

File: tests/drivers/dma/chan_blen_loop_transfer/boards/nucleo_f411re.conf [hunk 5]
```diff
@@ -0,0 +1 @@
+CONFIG_DMA=y
```
Comment (erwango) [inline tests/drivers/dma/chan_blen_loop_transfer/boards/nucleo_f411re.conf:1]: Can we move this commit to a dedicated PR ?

**Review by erwango — CHANGES_REQUESTED** [review]

**Comment by FRASTM** [comment]
spi/spi_loopback  example for nucleo_l476 and nucleo_f411 boards

**Comment by FRASTM** [comment]
decision to close this PR and to report all tests and configurations in the PR https://github.com/zephyrproject-rtos/zephyr/pull/23157

**Pull request closed without merging**